### PR TITLE
Cherry-pick commits for v1

### PR DESCRIPTION
**Pull request by aiqiaoy**
Cherry-pick commits for v1.2.0 release

Includes the following commits:

Fall back to gnu tar:
PR: https://github.com/actions/cache/pull/252
git cherry-pick 78809b91d73d12e094c66654de7b65727733dcc6..52046d14091e303ed4799d5c3ec3b96b9302023c

Adds socket timeout and validate file size
PR: https://github.com/actions/cache/pull/269
git cherry-pick f00dedfa6c708c51f4e4296f092458a4a891bddd..48b62c1c529d0cb84897a268c7c9a992f441b406

Better error handling during download 
PR: https://github.com/actions/cache/pull/284
git cherry-pick 1ed0c23029e0785d1732d054dc4aafb7cd760f23

Fix upload chunk retries
PR: https://github.com/actions/cache/pull/305
git cherry-pick 354f70a56cbf010d865e8bc8c5b6fb9d8e4f483c

Error handling for stream
PR: https://github.com/actions/cache/pull/300
git cherry-pick 4967c8e6c50fdd9eca6ebcc587a8fb80f4a169b3

Retries to all API
PR: https://github.com/actions/cache/pull/306
git cherry-pick 05b13411a0663afe7850fb8d87c8927b0bbb5f8e

Tested locally using `pipecli` and also ported Eric's e2e workflow change to get more coverage.

### CODE REVIEW

**Comment by aiqiaoy** [comment]
@joshmgross @dhadka There were a lot of merge conflicts so it would be great if both of you can review this PR.

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -4,51 +4,130 @@ on:
...
       run: npm run format-check
-
     - name: ESLint Check
       run: npm run lint
-
     - name: Build & Test
       run: npm run test
+
+  # End to end save and restore
+  test-save:
```
Comment (joshmgross) [inline .github/workflows/workflow.yml:52]: 🎉

File: .github/workflows/workflow.yml [hunk 2]
```diff
@@ -4,51 +4,130 @@ on:
   pull_request:
     branches:
       - master
+      - releases/**
```
Comment (joshmgross) [inline .github/workflows/workflow.yml:7]: We should add this path filter to the workflow in `master` as well.

**Comment by aiqiaoy** [comment]
Ran canary tests: https://github.com/actions/canary/actions?query=branch%3Aaiyan%2Fcache-v1-testing

**Pull request merged**